### PR TITLE
[SSADestroyHoisting] Handle access scopes.

### PR DESCRIPTION
**Pull request by nate-chandler**
Hoisting of lexical destroys has a different rule when the hoisting is done within an access:

> Mutable variable lifetimes that are formally modified in the middle of a lexical scope are anchored to the beginning of the lexical scope rather than to the end.

Here SSADestroyHoisting is modified to implement that rule.

While normally it checks for deinit barriers during hoisting, if the storage whose destroys are being hoisted is either a begin_access [modify] or an inout argument, it will not check for them.

Additionally, the pass is modified to enable additional optimizations around access scopes:

(1) For accesses to the address being destroyed, allow folding with the final storage use even if it occurs within an access scope so long as there are no deinit barriers between the end_access and that final storage use.

(2) For unrelated scopes, treat end_access as barriers iff there are no deinit barriers within the access scope.

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test

File: lib/SILOptimizer/Transforms/SSADestroyHoisting.cpp [hunk 1]
```diff
@@ -607,18 +609,23 @@ void SSADestroyHoisting::run() {
   // the end will result in hoisting inner begin_access' destroy_addrs first.
   while (!bais.empty()) {
     auto *bai = bais.pop_back_val();
-    changed |= hoistDestroys(bai, deleter);
+    changed |= hoistDestroys(bai, /*ignoreDeinitBarriers=*/false, deleter);
```
Comment (atrick) [inline lib/SILOptimizer/Transforms/SSADestroyHoisting.cpp:777]: a destroy within `begin_access [modify]` should ignore deinit barriers

File: lib/SILOptimizer/Transforms/SSADestroyHoisting.cpp [hunk 2]
```diff
@@ -234,20 +246,131 @@ class DeinitBarriers {
...
+      }
+    }
+
+    bool checkReachableBarrier(SILInstruction *inst) {
+      if (auto *eai = dyn_cast<EndAccessInst>(inst)) {
+        runningLiveAccessScopes.insert(eai->getBeginAccess());
+      } else if (auto *bai = dyn_cast<BeginAccessInst>(inst)) {
+        runningLiveAccessScopes.erase(bai);
+      }
+      bool isBarrier = result.isBarrier(inst);
```
Comment (atrick) [inline lib/SILOptimizer/Transforms/SSADestroyHoisting.cpp:345]: There is an implicit coupling between DestroyReacability::checkReachableBarrier and FindBarrierAccessScopes::checkReachableBarrier. Although it is nicely factored, I still think it's important to comment in both places that FindBarrierAccessScopes::checkReachableBarrier must return true for every instruction in which DestroyReacability::checkReachableBarrier returns true (except the end_access instructions which are being marked as barriers). This is required for correctness.

**Review by atrick — APPROVED** [review]
Fantastic!
- Now the logic for variable lifetime is well-defined and canonical across all forms of variables
- We have a simpler, more powerful optimization for removing copies
- We can remove a bunch of old, incomplete code
- This actually fixes some very difficult correctness issues with the previous destroy hoisting paths related to exclusivity

One of the optimizations is still disabled...

File: lib/SILOptimizer/Transforms/SSADestroyHoisting.cpp (hunk 2 above)
Comment (nate-chandler) [reply lib/SILOptimizer/Transforms/SSADestroyHoisting.cpp:350]: Thank you!  Added a comment to each describing the coupling of their implementations.

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
```
------- Performance (x86_64): -O -------

REGRESSION                                OLD    NEW    DELTA    RATIO
LessSubstringSubstring                    22     29     +31.8%   **0.76x (?)**
EqualSubstringSubstring                   22     29     +31.8%   **0.76x (?)**
EqualStringSubstring                      22     29     +31.8%   **0.76x (?)**
EqualSubstringString                      22     29     +31.8%   **0.76x (?)**
LessSubstringSubstringGenericComparable   22     29     +31.8%   **0.76x (?)**
EqualSubstringSubstringGenericEquatable   23     29     +26.1%   **0.79x (?)**
ObjectiveCBridgeStringHash                83     97     +16.9%   **0.86x (?)**
StringComparison_longSharedPrefix         295    333    +12.9%   **0.89x (?)**
SortStringsUnicode                        2010   2225   +10.7%   **0.90x (?)**
BridgeString.find.native.longNonASCII     393    433    +10.2%   **0.91x (?)**
RemoveWhereMoveInts                       20     22     +10.0%   **0.91x (?)**
StringHasPrefixAscii                      1430   1570   +9.8%    **0.91x (?)**

IMPROVEMENT                               OLD    NEW    DELTA    RATIO
FlattenListLoop                           1586   960    -39.5%   **1.65x (?)**
FlattenListFlatMap                        4254   3046   -28.4%   **1.40x (?)**
SortIntPyramid                            460    360    -21.7%   **1.28x**
MapReduceShort                            1250   1010   -19.2%   **1.24x (?)**
MapReduceClass2                           12     10     -16.7%   **1.20x (?)**
StringBuilder                             208    184    -11.5%   **1.13x (?)**
BucketSort                                88     78     -11.4%   **1.13x (?)**
StringBuilderSmallReservingCapacity       214    190    -11.2%   **1.13x (?)**
Data.hash.Empty                           52     47     -9.6%    **1.11x (?)**
Set.isSuperset.Seq.Int0                   119    108    -9.2%    **1.10x (?)**
StringUTF16Builder                        220    200    -9.1%    **1.10x (?)**
SortAdjacentIntPyramids                   1310   1195   -8.8%    **1.10x (?)**
DataAppendDataSmallToSmall                2720   2520   -7.4%    **1.08x (?)**

------- Code size: -O -------

IMPROVEMENT                    OLD     NEW     DELTA   RATIO
RomanNumbers.o                 6876    6588    -4.2%   **1.04x**
RangeOverlaps.o                6751    6543    -3.1%   **1.03x**
MapReduce.o                    28208   27616   -2.1%   **1.02x**
Combos.o                       5942    5822    -2.0%   **1.02x**
MonteCarloE.o                  2881    2825    -1.9%   **1.02x**
DictOfArraysToArrayOfDicts.o   22887   22471   -1.8%   **1.02x**
RangeAssignment.o              3331    3275    -1.7%   **1.02x**
BucketSort.o                   8679    8551    -1.5%   **1.01x**
FlattenList.o                  3857    3801    -1.5%   **1.01x**

------- Performance (x86_64): -Osize -------

REGRESSION                                               OLD    NEW    DELTA    RATIO
MapReduceLazyCollectionShort                             31     47     +51.6%   **0.66x (?)**
MapReduceLazySequence                                    44     66     +50.0%   **0.67x**
MapReduceLazyCollection                                  44     65     +47.7%   **0.68x**
EqualSubstringSubstringGenericEquatable                  22     30     +36.4%   **0.73x**
DropFirstAnySeqCRangeIterLazy                            107    143    +33.6%   **0.75x (?)**
DropFirstAnySeqCntRangeLazy                              107    143    +33.6%   **0.75x (?)**
LessSubstringSubstring                                   22     29     +31.8%   **0.76x (?)**
EqualStringSubstring                                     22     29     +31.8%   **0.76x (?)**
EqualSubstringString                                     22     29     +31.8%   **0.76x (?)**
LessSubstringSubstringGenericComparable                  22     29     +31.8%   **0.76x (?)**
DropFirstAnyCollection                                   85     112    +31.8%   **0.76x (?)**
EqualSubstringSubstring                                  23     30     +30.4%   **0.77x (?)**
SuffixAnyCollection                                      30     39     +30.0%   **0.77x (?)**
UTF8Decode_InitFromCustom_noncontiguous_ascii            623    809    +29.9%   **0.77x**
PrefixAnySeqCRangeIterLazy                               94     122    +29.8%   **0.77x**
PrefixAnySeqCntRangeLazy                                 94     122    +29.8%   **0.77x**
UTF8Decode_InitFromCustom_noncontiguous_ascii_as_ascii   718    904    +25.9%   **0.79x (?)**
PrefixWhileAnyCollectionLazy                             107    134    +25.2%   **0.80x (?)**
UTF8Decode_InitFromCustom_noncontiguous                  268    329    +22.8%   **0.81x (?)**
PrefixWhileAnyCollection                                 139    166    +19.4%   **0.84x (?)**
ObjectiveCBridgeStringHash                               83     97     +16.9%   **0.86x (?)**
DropWhileAnySeqCntRangeLazy                              127    148    +16.5%   **0.86x (?)**
DropWhileAnySeqCRangeIterLazy                            127    147    +15.7%   **0.86x (?)**
RemoveWhereSwapInts                                      20     23     +15.0%   **0.87x (?)**
DropFirstAnySeqCRangeIter                                98     111    +13.3%   **0.88x (?)**
DropFirstAnySeqCntRange                                  98     111    +13.3%   **0.88x (?)**
DropWhileAnyCollection                                   108    122    +13.0%   **0.89x (?)**
StringComparison_longSharedPrefix                        296    333    +12.5%   **0.89x (?)**
NormalizedIterator_fastPrenormal                         430    480    +11.6%   **0.90x (?)**
SortStringsUnicode                                       2015   2225   +10.4%   **0.91x (?)**
ObjectiveCBridgeStringGetASCIIContents                   314    346    +10.2%   **0.91x (?)**
BridgeString.find.native.longNonASCII                    393    433    +10.2%   **0.91x (?)**
RemoveWhereFilterInts                                    23     25     +8.7%    **0.92x (?)**
ArraySetElement                                          305    330    +8.2%    **0.92x (?)**
ObjectiveCBridgeStringCStringUsingEncoding               541    585    +8.1%    **0.92x (?)**
NormalizedIterator_latin1                                156    168    +7.7%    **0.93x (?)**
DropWhileAnyCollectionLazy                               146    157    +7.5%    **0.93x (?)**

IMPROVEMENT                                              OLD    NEW    DELTA    RATIO
FlattenListLoop                                          1584   959    -39.5%   **1.65x (?)**
ArrayAppendRepeatCol                                     670    500    -25.4%   **1.34x (?)**
ConvertFloatingPoint.MockFloat64Exactly2                 6      5      -16.7%   **1.20x (?)**
MapReduceShort                                           1250   1060   -15.2%   **1.18x (?)**
FlattenListFlatMap                                       2945   2531   -14.1%   **1.16x (?)**
StringBuilderSmallReservingCapacity                      213    189    -11.3%   **1.13x (?)**
StringBuilder                                            204    182    -10.8%   **1.12x (?)**
MapReduce                                                163    146    -10.4%   **1.12x (?)**
ProtocolDispatch                                         239    217    -9.2%    **1.10x (?)**
StringUTF16Builder                                       220    200    -9.1%    **1.10x (?)**
Set.isDisjoint.Seq.Int100                                123    113    -8.1%    **1.09x (?)**
StringInterpolationSmall                                 1150   1070   -7.0%    **1.07x (?)**
SetUnionInt25                                            76     71     -6.6%    **1.07x (?)**

------- Code size: -Osize -------

IMPROVEMENT                    OLD     NEW     DELTA   RATIO
RangeOverlaps.o                6414    6192    -3.5%   **1.04x**
RomanNumbers.o                 5220    5045    -3.4%   **1.03x**
MapReduce.o                    20957   20408   -2.6%   **1.03x**
DictOfArraysToArrayOfDicts.o   20293   19809   -2.4%   **1.02x**
MonteCarloE.o                  2845    2787    -2.0%   **1.02x**
RangeAssignment.o              3221    3163    -1.8%   **1.02x**
BucketSort.o                   8465    8323    -1.7%   **1.02x**
Hash.o                         18835   18520   -1.7%   **1.02x**
FlattenList.o                  3759    3701    -1.5%   **1.02x**
StringRemoveDupes.o            3760    3702    -1.5%   **1.02x**
Combos.o                       5965    5889    -1.3%   **1.01x**
CString.o                      5833    5775    -1.0%   **1.01x**

------- Performance (x86_64): -Onone -------

REGRESSION                                   OLD     NEW     DELTA    RATIO
ObjectiveCBridgeStringHash                   83      98      +18.1%   **0.85x (?)**
LessSubstringSubstringGenericComparable      52      59      +13.5%   **0.88x (?)**
EqualStringSubstring                         54      61      +13.0%   **0.89x (?)**
EqualSubstringString                         54      61      +13.0%   **0.89x (?)**
EqualSubstringSubstringGenericEquatable      53      59      +11.3%   **0.90x (?)**
RandomDouble01LCG                            21797   24213   +11.1%   **0.90x (?)**
LessSubstringSubstring                       55      61      +10.9%   **0.90x (?)**
EqualSubstringSubstring                      55      61      +10.9%   **0.90x (?)**
BridgeString.find.native.longNonASCII        393     433     +10.2%   **0.91x (?)**
ObjectiveCBridgeStringGetASCIIContents       316     346     +9.5%    **0.91x (?)**
SortStringsUnicode                           3295    3560    +8.0%    **0.93x (?)**
ObjectiveCBridgeStringCStringUsingEncoding   543     585     +7.7%    **0.93x (?)**

IMPROVEMENT                                  OLD     NEW     DELTA    RATIO
DictionarySwapAt                             2768    2512    -9.2%    **1.10x (?)**

------- Code size: -swiftlibs -------
```

**Pull request merged**